runners/direct-java/src/main/java/org/apache/beam/runners/direct/ExecutorServiceParallelExecutor.java

```diff
@@ -260,6 +260,11 @@ public State waitUntilFinish(Duration duration) throws Exception {
         }
       }
     }
+
+    if (Instant.now().isAfter(completionTime)) {
+      return null;
+    }
```

Comment:
This allows for a race condition where we exit the while loop above not due to a timeout but would turn it into a timeout here because of the timing.

We could clean-up the loop above so it only exits on timeout while all non-timeout returns happen within the loop with something like:
```
    while (Instant.now().isBefore(completionTime)) {
      // Get an update; don't block forever if another thread has handled it. The call to poll will
      // wait the entire timeout; this call primarily exists to relinquish any core.
      VisibleExecutorUpdate update = visibleUpdates.tryNext(Duration.millis(25L));
      if (pipelineState.get().isTerminal() || (update != null && isTerminalStateUpdate(update))) {
        // there are no updates to process and no updates will ever be published because the
        // executor is shutdown OR there has been an update and the update is terminal
        return pipelineState.get();
      } else if (update != null && update.thrown.isPresent()) {
        Throwable thrown = update.thrown.get();
        if (thrown instanceof Exception) {
          throw (Exception) thrown;
        } else if (thrown instanceof Error) {
          throw (Error) thrown;
        } else {
          throw new Exception("Unknown Type of Throwable", thrown);
        }
      }
    }
    return null;
```

Reply:
@lukecwik well spotted I totally overlooked it! Thanks I'll update this.

Reply:
Cleaned it up a little bit more and kept `(update == null && pipelineState.get().isTerminal())` to ensure that if an update has a `Throwable` it gets thrown. (keeps previous semantic).